### ISSUE TITLE
Add a new type Required

### ISSUE THREAD

**Issue by monolithed** [Microsoft/TypeScript#15012]
It would be nice to have a new type that allows property to be required as opposited to `Partial`:

```ts
interface X {
    x?: string 
}

Required<X>{ }; // Property 'x is missing in type '{}'
```

**Comment by masaeedu** [comment]
This will probably depend on #13253 or #12215, whichever ends up getting implemented. You need a mapped type where you subtract `undefined` and `null`.

**Comment by monolithed** [comment]
```ts
type Required<T> = T & {
  [P in keyof T]: T[P];
}

let a: Required<{ x: number }> = { }; // Property 'x is missing in type '{}'
```

it seems to work well.

**Comment by masaeedu** [comment]
@monolithed That snippet doesn't have any optional properties in the type you're passing as a type argument to `Required`. This corrected snippet doesn't have the expected error:

```ts
type Required<T> = T & {
  [P in keyof T]: T[P];
}

let a: Required<{ x?: number }> = { };
```

**Comment by monolithed** [comment]
@masaeedu, ah, you're right (

**Comment by aluanhaddad** [comment]
This is what you are looking for I believe. https://github.com/Microsoft/TypeScript/issues/13224

**Comment by monolithed** [comment]
> This is what you are looking for I believe. #13224

@aluanhaddad, So, why the following operation `[P in keyof T]: T[P]` is not working as expected?

**Comment by PyroVortex** [comment]
@monolithed 
The operation is a homomorphic mapping, which have been configured to preserve optional and readonly modifiers. To get the desired behavior, you would need to do:
```ts
type RequiredInternal<T, K extends keyof T> = { [P in K]: T[P] }; // Use type parameter to force it not to be considered homomorphic
type Required<T> = T & RequiredInternal<T, keyof T>;

let a: Required<{ x?: number }> = { }; // => Property 'x' is missing in type '{}'
```

**Comment by masaeedu** [comment]
@PyroVortex No, I tried that yesterday (I think it might have been your comment on another issue). With `tsc` version `2.2.1`, and the following files:

```json
// tsconfig.json
{
    "compilerOptions": {
        "module": "commonjs",
        "target": "es5",
        "strictNullChecks": true,
        "noImplicitAny": false,
        "sourceMap": false
    }
}
```

```ts
// main.ts
type RequiredInternal<T, K extends keyof T> = { [P in K]: T[P] }; // Use type parameter to force it not to be considered homomorphic
type Required<T> = T & RequiredInternal<T, keyof T>;

let a: Required<{ x?: number }> = { }; // No error here
let x: number = a.x;                            // Error here because number | undefined is not assignable to number
```

The expected error does not appear, and `a` is typed as having a potentially undefined `x` property (`number | undefined`).

**Comment by PyroVortex** [comment]
@masaeedu 
Huh. In the current version in the TypeScript Playground with strictNullChecks enabled, I get an error for the first assignment (missing property), but I do still see the error you are mentioning for the second.

In general, we lack an ability to remove a type from a union using pure type manipulation (we can do it with an actual value).

**Comment by masaeedu** [comment]
@PyroVortex How did you enable `strictNullChecks` in playground?

**Comment by aluanhaddad** [comment]
![image](https://cloud.githubusercontent.com/assets/5034656/24737377/23a74702-1a5c-11e7-82e0-5be0ebb065b3.png)

**Comment by priithaamer** [comment]
@PyroVortex @masaeedu I am too trying to create type that maps optional fields to required fields. I get the example above to raise an error with TS 2.1.4 and in the playground (which it seems to be using). From 2.1.5 and up it does not give me an error anymore. Any idea why it is so?

**Comment by thehappycoder** [comment]
This feature is very handy in react stateful components when you want to specify the default state, and you want to make sure you don't miss any state.

**Comment by hill0826** [comment]
I'm still running into issues with this in TS 2.4.1. I've tried breaking the homomorphic relationship, as [described](https://github.com/Microsoft/TypeScript/issues/13723#issuecomment-275730246) in #13723, and I believe this does that because the fields are now required but instead of removing all traces of optional parameters it changes them to required fields with a union type `undefined`. Is this intentional?

```ts
type RequiredInternal<T extends {[key: string]: any}, K extends string> = { [P in K]: T[P]};
type Required<T> = RequiredInternal<T, keyof T>;

// As expected: No compile error
const val1: Required<{x?: number}> = {x: 1};

// As expected: Compile error:
// TS2322: Type '{}' is not assignable to type 'RequiredInternal<{ x?: number | undefined; }, "x">'. Property 'x' is missing in type '{}'.
const val2: Required<{x?: number}> = {};

// Not expected?: No compile error
// Expected something like TS2322: Type '{ x: undefined; }' is not assignable to type 'RequiredInternal<{ x?: number | undefined; }, "x">'. Types of property 'x' are incompatible. Type 'undefined' is not assignable to type 'number'.
const val3: Required<{x?: number}> = {x: undefined};
```

**Comment by MeirionHughes** [comment]
@PyroVortex I'm afraid even with strictNullCheck, your example still results in a being optional (typescript 2.4): 

![image](https://user-images.githubusercontent.com/3584509/31310560-ea63aa5e-ab91-11e7-9ca3-e7bd10b4a4d8.png)

**Comment by ssonne** [comment]
Try this:

```
// Non-homomorphic keys
type NHKeys<T> = ({[P in keyof T]: P } & { [x:string]:never })[keyof T];

type Required<T> = {
  [K in NHKeys<T>]:T[K];
};

interface Foo {
  a?:string;
  b:string;
};

const foo:Foo = {b:'abc'};
const fooRequired:Required<Foo> = {b:'abc'}; // error
```

**Comment by andy-ms** [comment]
@sandersn @weswigham Is there a better way to do this?

**Comment by weswigham** [comment]
I want to say... no? @ssonne seems to have it. The core is breaking the inference of a homomorphic relationship within the mapped type; and it seems like we keep that for as long as we can (traversing aliases and simplifying unions to look for it). The intersection-index indirection seems like a sound way to go.

**Comment by MeirionHughes** [comment]
@monolithed why don't you make PR --or better yet, invite @ssonne to do it because he figured it out-- to add `Required<>` to the [types bundled by typescript itself ](https://github.com/Microsoft/TypeScript/blob/a6555921f4b1d51281b1241c69e91e5b82d2f07e/lib/lib.es5.d.ts#L1313-L1329)

**Comment by monolithed** [comment]
@MeirionHughes https://github.com/Microsoft/TypeScript/pull/19398

**Comment by monolithed** [comment]
The PR is closed due to:

> I do not think this is a type we want to include in the standard library. if we were to address #15012, we need a type operator/ an addition to mapped types that allows such transformation.
--

**Comment by DanielRosenwasser** [comment]
@ssonne I think this is a bug, and I have a fix available for it... so... I'm sorry?

**Comment by pelotom** [comment]
@DanielRosenwasser So you’re saying this `Required` trick is about to be “fixed” out of existence? 😕

**Comment by hill0826** [comment]
I use one of the variants of the aforementioned `Required` type definition in my project, and it works for its purpose (requiring all fields to be set), but when I'm using this what I usually really want is to ensure that all fields are non-null and non-undefined. `Required` doesn't help us there since it still allows the field to be set as `undefined`.

Thankfully, there are a few tickets out for "subtraction" types, which might solve this: #4183, #12215, #15059.

**Comment by Taytay** [comment]
I just want to get clarification on this @DanielRosenwasser. Is the fact that this type works a quirk of a bug you think that @ssonne has found? So if we were to adopt this, it would stop working once the bug is fixed?

**Comment by nevir** [comment]
The workaround @ssonne has doesn't work in all contexts, unfortunately, too.  For example, if you enable `strictNullChecks`:

```ts
interface Optional {
  foo?: () => {}
}
type Strict = Required<Optional>

const impl: Strict = {
  foo() {}
};

impl.foo();
// ^ Object is possibly 'undefined'.
```
[[Playground](https://www.typescriptlang.org/play/index.html#src=type%20NHKeys%3CT%3E%20%3D%20(%7B%5BP%20in%20keyof%20T%5D%3A%20P%20%7D%20%26%20%7B%20%5Bx%3Astring%5D%3Anever%20%7D)%5Bkeyof%20T%5D%3B%0D%0A%0D%0Atype%20Required%3CT%3E%20%3D%20%7B%0D%0A%20%20%5BK%20in%20NHKeys%3CT%3E%5D%3AT%5BK%5D%3B%0D%0A%7D%3B%0D%0A%0D%0Ainterface%20Optional%20%7B%0D%0A%20%20foo%3F%3A%20()%20%3D%3E%20void%0D%0A%7D%0D%0Atype%20Strict%20%3D%20Required%3COptional%3E%0D%0A%0D%0Aconst%20impl%3A%20Strict%20%3D%20%7B%0D%0A%20%20foo()%20%7B%7D%0D%0A%7D%3B%0D%0A%0D%0Aimpl.foo()%3B%0D%0A%2F%2F%20%5E%20Object%20is%20possibly%20'undefined'.%0D%0A)]

**Comment by falsandtru** [comment]
If TypeScript will support Required types, I propose the following new syntax.

```ts
type Required<T> = {
  [P in keyof T]!: T[P];
};

**Comment by falsandtru** [comment]
We can make required properties, but they still have nullable values. The last piece is NonNullable type.

```ts
type Required<T> = {
  [P in Purify<keyof T>]: NonNullable<T[P]>;
};
type Purify<T extends string> = { [P in T]: T; }[T];
type NonNullable<T> = T - undefined - null;
```

**Comment by mpawelski** [comment]
Well, right now we can have some workaround that allows you to define required type from partial one
```ts
type PartialPerson = {
  name? : number;
  surname? : string;
}

declare function makeRequired<T>(notRequired : Partial<T>): T;

let __requiredPersonVariable = null as any && makeRequired({} as PartialPerson);
let RequiredPerson = typeof __requiredPersonVariable;   
```

Then if we got #6606 we could do something like this (I _guess_, I'm not sure how it would work with generic type aliases if it was implemented):
```ts
type Required<T> = typeof( null as any && makeRequired({} as T))
```

For me It's another example how useful #6606 would be if we had it :)

**Comment by pelotom** [comment]
@mpawelski minor correction:

```ts
type RequiredPerson = typeof __requiredPersonVariable;
```

**Comment by niieani** [comment]
I've found out that `T & {}` simplifies to non-nullable type, i.e.:

```ts
type Required<T> = {
  [P in Purify<keyof T>]: NonNullable<T[P]>;
};
type Purify<T extends string> = { [P in T]: T; }[T];
type NonNullable<T> = T & {};
```

Thanks to @falsandtru for the Purify method.

[Playground](https://www.typescriptlang.org/play/#src=type%20Required%3CT%3E%20%3D%20%7B%0D%0A%20%20%5BP%20in%20Purify%3Ckeyof%20T%3E%5D%3A%20NonNullable%3CT%5BP%5D%3E%3B%0D%0A%7D%3B%0D%0Atype%20Purify%3CT%20extends%20string%3E%20%3D%20%7B%5BP%20in%20T%5D%3A%20T%3B%20%7D%5BT%5D%3B%0D%0Atype%20NonNullable%3CT%3E%20%3D%20T%20%26%20%7B%7D%3B%0D%0A%0D%0Ainterface%20Example1%20%7B%0D%0A%20%20one%3F%3A%20string%0D%0A%7D%0D%0Ainterface%20Example2%20%7B%0D%0A%20%20two%3A%20string%20%7C%20null%0D%0A%7D%0D%0Ainterface%20Example3%20%7B%0D%0A%20%20three%3A%20string%20%7C%20undefined%0D%0A%7D%0D%0Ainterface%20Example4%20%7B%0D%0A%20%20four%3A%20string%20%7C%20null%20%7C%20undefined%0D%0A%7D%0D%0Ainterface%20Example5%20%7B%0D%0A%20%20five%3F%3A%20string%20%7C%20null%20%7C%20undefined%0D%0A%7D%0D%0A%0D%0A%2F%2F%20should%20fail%3A%0D%0Aconst%20testCase1%3A%20Required%3CExample1%3E%20%3D%20%7B%0D%0A%20%20one%3A%20undefined%2C%0D%0A%7D%0D%0A%0D%0Aconst%20testCase2%3A%20Required%3CExample2%3E%20%3D%20%7B%0D%0A%20%20two%3A%20null%2C%0D%0A%7D%0D%0A%0D%0Aconst%20testCase3%3A%20Required%3CExample3%3E%20%3D%20%7B%0D%0A%20%20three%3A%20undefined%2C%0D%0A%7D%0D%0A%0D%0Aconst%20testCase4%3A%20Required%3CExample4%3E%20%3D%20%7B%0D%0A%20%20four%3A%20undefined%2C%0D%0A%7D%0D%0A%0D%0Aconst%20testCase5%3A%20Required%3CExample5%3E%20%3D%20%7B%0D%0A%20%20five%3A%20undefined%2C%0D%0A%7D%0D%0A%0D%0Aconst%20testCase1b%3A%20Required%3CExample1%3E%20%3D%20%7B%0D%0A%20%20one%3A%201%2C%0D%0A%7D%0D%0A%0D%0Aconst%20testCase2b%3A%20Required%3CExample2%3E%20%3D%20%7B%0D%0A%20%20two%3A%201%2C%0D%0A%7D%0D%0A%0D%0Aconst%20testCase3b%3A%20Required%3CExample3%3E%20%3D%20%7B%0D%0A%20%20three%3A%201%2C%0D%0A%7D%0D%0A%0D%0Aconst%20testCase4b%3A%20Required%3CExample4%3E%20%3D%20%7B%0D%0A%20%20four%3A%201%2C%0D%0A%7D%0D%0A%0D%0Aconst%20testCase5b%3A%20Required%3CExample5%3E%20%3D%20%7B%0D%0A%20%20five%3A%201%2C%0D%0A%7D%0D%0A%0D%0A%2F%2F%20should%20pass%3A%0D%0Aconst%20testCase1a%3A%20Required%3CExample1%3E%20%3D%20%7B%0D%0A%20%20one%3A%20'ok'%2C%0D%0A%7D%0D%0A%0D%0Aconst%20testCase2a%3A%20Required%3CExample2%3E%20%3D%20%7B%0D%0A%20%20two%3A%20'ok'%2C%0D%0A%7D%0D%0A%0D%0Aconst%20testCase3a%3A%20Required%3CExample3%3E%20%3D%20%7B%0D%0A%20%20three%3A%20'ok'%2C%0D%0A%7D%0D%0A%0D%0Aconst%20testCase4a%3A%20Required%3CExample4%3E%20%3D%20%7B%0D%0A%20%20four%3A%20'ok'%2C%0D%0A%7D%0D%0A%0D%0Aconst%20testCase5a%3A%20Required%3CExample5%3E%20%3D%20%7B%0D%0A%20%20five%3A%20'ok'%2C%0D%0A%7D) (remember to enable strictNullChecks).

**Comment by falsandtru** [comment]
Great! @weswigham @sandersn @ahejlsberg Can we use `type NonNullable<T> = T & {};` as a stable API?

**Comment by pelotom** [comment]
Very nice @niieani! And it makes sense:

```
NonNullable<A | undefined> = (A | undefined) & {}
                           = A & {} | undefined & {} // & distributes over |
                           = A & {} | never          // undefined and {} are disjoint
                           = A & {}                  // never is an identity to |
```

At that point either `A` is not possibly `undefined` or `null`, in which case `A & {} = A`, or it is, in which case you repeat the above process until `undefined` and `null` have been stripped out of the type.

**Comment by pelotom** [comment]
For those interested, I've added `Required` (and `NonNullable`) to [Type Zoo](https://github.com/pelotom/type-zoo).

**Comment by falsandtru** [comment]
Now the latest version is:

```ts
export type DeepRequired<T, U extends object | undefined = undefined> =
  T extends object
    ? { [P in Purify<keyof T>]: NonNullable<T[P]> extends NonNullable<U | Function> ? NonNullable<T[P]> : DeepRequired<NonNullable<T[P]>, U>; }
    : T;
```

https://github.com/falsandtru/spica/blob/master/src/lib/type.ts
https://github.com/falsandtru/spica/blob/master/src/lib/type.test.ts

**Comment by weswigham** [comment]
@falsandtru And as of today there's no need for `Purify`:
```ts
type Required<T> =
  T extends object
    ? { [P in keyof T]-?: NonNullable<T[P]>; }
    : T;
type DeepRequired<T, U extends object | undefined = undefined> =
  T extends object
    ? { [P in keyof T]-?: NonNullable<T[P]> extends NonNullable<U | Function | Class> ? NonNullable<T[P]> : DeepRequired<NonNullable<T[P]>, U>; }
    : T;
```
this should also handle assignability while things are still generic better, too; or at least that's the goal.

**Comment by falsandtru** [comment]
Right, I'll try to use the new syntax later.
